### DISCUSSION TITLE
关于张量程序的转换

### DISCUSSION THREAD

**Discussion by LucienXian** [mlc-ai/mlc-zh#14] [Q&A] [answered]
课程中的例子都是需要手动指定张量程序的转换行为，比如loop split，reorder等。真实的机器学习编译也是这样的吗，也是需要手动指明index之间有没有依赖，如何做转换（比如是拆成3个子loop），人工地提供信息，还是会通过程序去分析。这里是不是跟传统的编译类似，比如我指明一个函数是inline的，但编译器不一定真正去做内联，人工提供的信息只是一种建议。还是说人工提供的信息类似一种编译选项。

**Comment by jinhongyii** [comment]
真实的机器学习编译会通过程序去分析需要执行的张量程序转换，当程序分析到一个函数需要inline的时候，它就会执行。

**Comment by Hzfengsy** [comment, marked as answer]
感谢你的提问，你的提问中涉及了两个概念：如何做变换，和什么时候需要变换。

loop split，reorder都是在讲我们怎么做变换，相当于是从一个程序（IR）变换到另一个程序。与之对应的是编译器中的inline变换（而不是用户建议），即把两个function的IR变换成一个function的IR。

第二个方面是什么时候做变换，即我们什么时候去做变换，给变换的参数是什么（比如index之间有没有依赖，比如传统编译器中我们是不是“真正”去inline）。

这两层概念在ML和传统编译器领域都是很重要的，只是涉及面不同。前者更加面相与编译器开发者，后者更多的会与使用者相关。本课程主要涉及的是前者，即MLC是怎么做的；与之对应的MLC是怎么用的才是后者更关心的问题。

**Reply by hbsun2113** [reply]
@Hzfengsy 谢谢解答！
所以本课程着重讲授tvm的原理，而不是tvm的使用方法，对吗？
如果我个人想着重了解tvm的使用，该从哪里入手呢？比如我想了解`spatial`的使用办法，似乎在官方文档里就没有看到很好的解释：
<img width="1271" alt="image" src="https://user-images.githubusercontent.com/27994442/175808001-a5b8632c-1cf4-412b-a14a-b5de5c74c2da.png">

**Reply by Hzfengsy** [reply]
目前官方文档是不完善的，社区正在补充和扩展。如果了解tvm的使用，可以阅读user guild内部相关内容。
